includes/class-wc-form-handler.php

```diff
@@ -904,6 +906,9 @@ private static function add_to_cart_handler_variable( $product_id ) {
 				}
 			}
 
+			// Merge variation attributes and posted attributes.
+			$posted_and_variation_attributes = array_merge( $posted_attributes, $variation_attributes );
```

Comment:
I'm not sure about the order of arrays here. 

In case all attributes are specified for a variation, then it makes sense to leave it as it is, since `$posted_attributes` could be (theoretically) incorrect.

On the other hand, there are variations that can have an attribute value set to `Any`. 
<img width="1421" alt="Screenshot 2020-04-22 at 13 34 11" src="https://user-images.githubusercontent.com/2207451/79977089-f72ee900-849d-11ea-8994-ebef0d4873ba.png">


Then `$posted_attributes` can theoretically further specify which option customer chose, e.g. via args `?add-to-cart=448&attribute_color=Black`. In that case, the extra attribute value would get lost, no?

Reply:
It doesn't appear the extra attribute value gets lost, because `$adding_to_cart->get_variation_attributes()` only returns attributes that are defined for the variation, in your screenshot, the size attribute. If color was specified via args, the `array_merge` should merge size and color keys into a single array, regardless of order.

I believe that the only case of something being lost would be if you specified size as an argument. For example, in your screenshot `?add-to-cart=448&attribute_pa_size=s&attribute_pa_color=green`, the `$posted_attributes` size would be lost and replaced with the attribute size from `$variation_attributes`. This would add the medium green shirt to the cart, ignoring `attribute_pa_size` in the args.

However, when I try to do that with a variation product id, for any attribute value other than the one belonging to the variation, I receive this error if I swap the order of the `array_merge` arguments (example `?add-to-cart=448&attribute_pa_size=s&attribute_pa_color=green`)

![image](https://user-images.githubusercontent.com/363749/80136149-f01fdd80-8566-11ea-98e8-1431b668ae77.png)

I think it's a matter of preference whether or not we display that error messages when you try to override an attribute belonging to the variation. Do you have a preference?

Reply:
I ended up swapping the order and adding a check to test for the error message when attempting to override an attribute belonging to the variation.

Reply:
> because $adding_to_cart->get_variation_attributes() only returns attributes that are defined for the variation,

For me, `$adding_to_cart->get_variation_attributes()` returns empty string for unassigned variation:
<img width="1465" alt="Screenshot 2020-04-28 at 20 29 07" src="https://user-images.githubusercontent.com/2207451/80523911-1c849100-898f-11ea-94e1-7f325e8056d0.png">

so I thought `array_merge($posted, $variation)` would replace `$posted` attribute with the empty string from `$variation`, e.g. when testing:

```
php > $posted = array('size' => 'm', 'color' => 'blue');
php > $variation = array('size' => '', 'color' => 'blue');
php > var_dump(array_merge($posted, $variation));
array(2) {
  ["size"]=>
  string(0) ""
  ["color"]=>
  string(4) "blue"
}
```

Thanks for updating, I think it behaves better now.